### PR TITLE
Fix #5164 by overriding environment variables when new-building custom Setups

### LINKED ISSUE #5164: Cannot install pango with new-build when using local copy of gtk2hs-buildtools
I'm experiencing this issue with both:

```
$ cabal --version
cabal-install version 2.0.0.1
compiled using version 2.0.1.1 of the Cabal library 
$ cabal-head --version
cabal-install version 2.1.0.0
compiled using version 2.1.0.0 of the Cabal library
```

To trigger the issue:

```
$ cabal get pango-0.13.4.0
Unpacking to pango-0.13.4.0/
$ cd pango-0.13.4.0/
$ cabal get gtk2hs-buildtools-0.13.3.1
Unpacking to gtk2hs-buildtools-0.13.3.1/
$ cat > cabal.project << EOF
> packages: .
>           ./gtk2hs-buildtools-0.13.3.1
> EOF
$ cabal new-build all
Resolving dependencies...
In order, the following will be built (use -v for more details):
 - gtk2hs-buildtools-0.13.3.1 (lib) (first run)
 - gtk2hs-buildtools-0.13.3.1 (exe:gtk2hsTypeGen) (first run)
 - gtk2hs-buildtools-0.13.3.1 (exe:gtk2hsHookGenerator) (first run)
 - gtk2hs-buildtools-0.13.3.1 (exe:gtk2hsC2hs) (first run)
 - glib-0.13.5.0 (lib:glib) (requires build)
 - cairo-0.13.4.2 (lib:cairo) (requires build)
 - pango-0.13.4.0 (lib:pango) (first run)
Configuring library for gtk2hs-buildtools-0.13.3.1..
Preprocessing library for gtk2hs-buildtools-0.13.3.1..
<build output elided>
[1 of 1] Compiling Main             ( /home/rgscott/Documents/Hacking/Haskell/pango-0.13.4.0/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pango-0.13.4.0/setup/setup.hs, /home/rgscott/Documents/Hacking/Haskell/pango-0.13.4.0/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pango-0.13.4.0/setup/Main.o )

/home/rgscott/Documents/Hacking/Haskell/pango-0.13.4.0/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pango-0.13.4.0/setup/setup.hs:51:18: warning: [-Wdeprecations]
    In the use of ‘autogenModulesDir’
    (imported from Distribution.Simple.BuildPaths):
    Deprecated: "If you can, use 'autogenComponentModulesDir' instead, but if you really wanted package-global generated modules, use 'autogenPackageModulesDir'.  In Cabal 2.0, we avoid using autogenerated files which apply to all components, because the information you often want in these files, e.g., dependency information, is best specified per component, so that reconfiguring a different component (e.g., enabling tests) doesn't force the entire to be rebuilt.  'autogenPackageModulesDir' still provides a place to put files you want to apply to the entire package, but most users of 'autogenModulesDir' should seriously consider 'autogenComponentModulesDir' if you really wanted the module to apply to one component."
   |
51 |     targetDir  = autogenModulesDir lbi
   |                  ^^^^^^^^^^^^^^^^^

/home/rgscott/Documents/Hacking/Haskell/pango-0.13.4.0/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pango-0.13.4.0/setup/setup.hs:55:3: warning: [-Wdeprecations]
    In the use of ‘die’ (imported from Distribution.Simple.Utils):
    Deprecated: "Messages thrown with die can't be controlled with Verbosity; use die' instead, or dieNoVerbosity if Verbosity truly is not available"
   |
55 |   die $ "unexpected pango version number: " ++ display version
   |   ^^^
Linking /home/rgscott/Documents/Hacking/Haskell/pango-0.13.4.0/dist-newstyle/build/x86_64-linux/ghc-8.2.2/pango-0.13.4.0/setup/setup ...
Configuring pango-0.13.4.0...
/u/rgscott/.cabal/share/x86_64-linux-ghc-8.2.2/gtk2hs-buildtools-0.13.3.1/hierarchyGen/Hierarchy.chs.template: openFile: does not exist (No such file or directory)cabal: Failed to build pango-0.13.4.0. The failure occurred during the
configure step.
```

Note that this bug will not be triggered if you have installed `gtk2hs-buildtools-0.13.3.1` previously using `cabal install`, as that will cause `~/.cabal/share/x86_64-linux-ghc-8.2.2/gtk2hs-buildtools-0.13.3.1/hierarchyGen/Hierarchy.chs.template` to have been created.

Moreover, this bug does _not_ occur if `gtk2hs-buildtools-0.13.3.1` is not put into `cabal.project` and is instead downloaded from Hackage.

I suspect the issue may have something to do with [this line of code in `gtk2hs-buildtools-0.13.3.1`](http://hackage.haskell.org/package/gtk2hs-buildtools-0.13.3.1/docs/src/TypeGen.html#line-148):

```haskell
templateFile <- getDataFileName "hierarchyGen/Hierarchy.chs.template"
```

### PR DESCRIPTION
**Pull request by RyanGlScott**
This is accomplished by adding a `useExtraEnvOverrides` field to `SetupScriptOptions`, which allows for setting (or unsetting) any environment variables you wish. This is used when invoking a custom `Setup` script executable to set the values of variables like `<component>_datadir` when using `new-build`. (Not doing this before led to #5164).

Along the way, I moved `dataDirsEnvironmentForPlan` into `Distribution.Client.ProjectPlanning.Types` (as that seemed like the most logical place to put it) so that it could be used in `Distribution.Client.ProjectPlanning` as well, where we also need to initialize `useExtraEnvOverrides`.

I attempted to add a test, but I failed miserably at actually running the test suite, so I have no idea if it actually works or not. Fingers crossed.

Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [ ] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: cabal-install/changelog [hunk 1]
```diff
@@ -66,6 +66,8 @@
 	* Paths_ autogen modules now compile when `RebindableSyntax` or
 	`OverloadedStrings` is used in `default-extensions`.
 	[stack#3789](https://github.com/commercialhaskell/stack/issues/3789)
+    * getDataDir` and other `Paths_autogen` functions now work correctly
```
Comment (RyanGlScott) [inline cabal-install/changelog]: I put this in the `2.2.0.0` release notes, as that was the most recent one listed. If this won't make it until the next release, let me know and I can make a section for `2.x.0.0` (where `x` is presumably 4).

File: cabal-install/Distribution/Client/SetupWrapper.hs [hunk 2]
```diff
@@ -414,7 +420,8 @@ internalSetupMethod verbosity options bt args = do
   inDir (useWorkingDir options) $ do
     withEnv "HASKELL_DIST_DIR" (useDistPref options) $
       withExtraPathEnv (useExtraPathEnv options) $
-        buildTypeAction bt args
+        withEnvOverrides (useExtraEnvOverrides options) $
```
Comment (RyanGlScott) [inline cabal-install/Distribution/Client/SetupWrapper.hs]: Having to explicitly set environment variables here feels somewhat strange (see https://github.com/haskell/cabal/issues/5164#issuecomment-372447822). But then again, we also do this for `HASKELL_DIST_DIR` and `PATH` above, so I opted to do the same here for consistency's sake.

**Comment by RyanGlScott** [comment]
@hvr helped me realize why the tests were failing... I forgot to include `other-modules: Paths_...`. D'oh.

Let's see if it passes with that change.

**Comment by hvr** [comment]
@RyanGlScott there's still a few failures w/ the new test-case, e.g.

```
Actual output differs from expected:
stderr:
--- /home/travis/build/haskell/cabal/cabal-testsuite/PackageTests/NewBuild/T5164/cabal.dist/cabal.out.normalized	2018-03-14 01:24:03.454152691 +0000
+++ /home/travis/build/haskell/cabal/cabal-testsuite/PackageTests/NewBuild/T5164/cabal.dist/cabal.comp.out.normalized	2018-03-14 01:24:03.454152691 +0000
@@ -0,0 +1,13 @@
+# cabal new-build
+Resolving dependencies...
+Build profile: -w ghc-<GHCVER> -O1
+In order, the following will be built:
+ - setup-lib-1.0 (lib) (first run)
+ - uses-custom-setup-1.0 (lib:uses-custom-setup) (first run)
+Configuring library for setup-lib-1.0..
+Preprocessing library for setup-lib-1.0..
+Building library for setup-lib-1.0..
+Configuring uses-custom-setup-1.0...
+Preprocessing library for uses-custom-setup-1.0..
+Building library for uses-custom-setup-1.0..
+List of package sources written to file '/tmp/cabal-list-sources.-30948/cabal-sdist-list-sources'
*** unexpected failure for PackageTests/NewBuild/T5164/cabal.test.hs
```

**Comment by RyanGlScott** [comment]
I'm not sure how to trigger this error, as the test appears to be working locally for me:

```
~/.../Haskell/cabal/cabal-testsuite on T5164
$ ../dist-newstyle/build/x86_64-linux/ghc-8.2.2/cabal-testsuite-2.3.0.0/build/cabal-tests/cabal-tests PackageTests/NewBuild/CmdRun/ExitCodePropagation/cabal.test.hs --with-cabal ../dist-newstyle/build/x86_64-linux/ghc-8.2.2/cabal-install-2.3.0.0/build/cabal/cabal
<elided wall of text>
OK
~/.../Haskell/cabal/cabal-testsuite on T5164
$ echo $?
0
```

**Comment by RyanGlScott** [comment]
Hooray, the tests seem to be passing now! One of the [OSX builds seemed to stall out](https://travis-ci.org/haskell/cabal/jobs/353353071#L1707), but I don't think this is my doing.

**Comment by hvr** [comment]
@RyanGlScott yeah, the OSX ones often run into timeouts or other weirdness; I consider CI green

**Comment by RyanGlScott** [comment]
Gentle ping. Is this in a mergeable state, or is there more to be done here?

**Comment by 23Skidoo** [comment]
Mergeable IMO. I'd like to backport this to 2.2 as well.

**Comment by 23Skidoo** [comment]
OK, I'm merging this since it was green before and the failures are caused by the hackage-security update.

**Comment by RyanGlScott** [comment]
Awesome, thanks!

**Comment by 23Skidoo** [comment]
Also cherry-picked into 2.2.

**Pull request merged**